state/protocol/badger/mutator.go

```diff
@@ -1048,16 +1044,17 @@ func (m *FollowerState) MarkValid(blockID flow.Identifier) error {
 			return fmt.Errorf("could not insert validity for block (id=%x, height=%d): %w", blockID, header.Height, err)
 		}
 
-		// trigger BlockProcessable for parent blocks above root height
-		if parent.Height > rootHeight {
-			// TODO deliver protocol events async https://github.com/dapperlabs/flow-go/issues/6317
-			m.consumer.BlockProcessable(parent)
-		}
 		return nil
 	})
 	if err != nil {
 		return fmt.Errorf("could not mark block as valid (%x): %w", blockID, err)
 	}
 
+	// skip emitting BlockProcessable for root block, as it is always processable
+	if parent.Height > rootHeight {
+		// emit protocol events after database transaction succeeds
+		m.consumer.BlockProcessable(parent)
+	}
+
```

Comment:
⚠️ I think this might break the guarantee that the order of events to be the consistent with the order of these finalized blocks being saved to database.

Consider `Extend` method is called concurrently for 3 blocks: `A, B, C`, and because of database transaction, they are stored into database in this order `A, B, C`.

After they are stored, `BlockProcessable` will be called outside of the db transaction and in it's own goroutine concurrently, it's possible that a consumer will receive `BlockProcessable` event in `B, A, C`, which is different from the order they are saved in db. 

Should we add a mutex lock to the `Extend` method? But that would disallow concurrent processing of blocks on different fork.

Reply:
I don't think we had this guarantee before, either.

Badger guarantees that database state changes are persisted in an ACID-compliant way, it doesn't guarantee anything about the execution order of code which operates on a `*badger.Transaction` object. 

In general, we haven't used or tested concurrent block insertion/validation with the `MutatableState` that I'm aware of. Since this PR is not changing the behaviour, I'm inclined to leave any other changes to `Extend` until we make changes to the `compliance` engine to make use of concurrent block validation.